File: Sources/CountriesFlagsObjC/source/Countries/UIImage+SovietUnion.m
//
//  UIImage+SovietUnion.m
//  CountriesFlags
//

#import "Countries/UIImage+SovietUnion.h"

#import "Extensions/UIBezierPath+Star.h"

@implementation UIImage (SovietUnion)

+ (UIImage * _Nullable)flagSovietUnionWithSize:(CGSize)size {
    UIColor* colorRed = [UIColor colorWithRed:203.0f/255.0f
                                        green:0.0f
                                         blue:0.0f
                                        alpha:1.0f];
    UIColor* colorYellow = [UIColor colorWithRed:1.0f
                                           green:215.0f/255.0f
                                            blue:0.0f
                                           alpha:1.0f];
    
    UIGraphicsBeginImageContextWithOptions(size, YES, 0.0f);
    
    [colorRed setFill];
    UIRectFill(CGRectMake(0.0f, 0.0f, size.width, size.height));
    
    [colorYellow setFill];
    // TODO: rework star
    [colorYellow setFill];
    CGPoint starCenter = CGPointMake(size.width / 100.0f * 16.7f, size.height / 100.0f * 12.5f);
    [[UIBezierPath starAtPoint:starCenter
                        radius:size.height / 16.0
                          type:FlagStarTypeFivePointed] fill];
    [colorRed setFill];
    [[UIBezierPath starAtPoint:starCenter
                        radius:size.height / 24.5
                          type:FlagStarTypeFivePointed] fill];

    [colorYellow setFill];
    // draw hammer
    UIBezierPath* path = [UIBezierPath bezierPath];
    [path moveToPoint:CGPointMake(size.width / 1000.0f * 116.0f, size.height / 1000.0f * 290.0f)];
    [path addLineToPoint:CGPointMake(size.width / 1000.0f * 148.0f, size.height / 1000.0f * 227.0f)];
    [path addLineToPoint:CGPointMake(size.width / 1000.0f * 175.0f, size.height / 1000.0f * 234.0f)];
    [path addLineToPoint:CGPointMake(size.width / 1000.0f * 155.3f, size.height / 1000.0f * 273.0f)];
    [path addLineToPoint:CGPointMake(size.width / 1000.0f * 225.0f, size.height / 1000.0f * 406.0f)];
    [path addArcWithCenter:CGPointMake(size.width / 1000.0f * 221.0f, size.height / 1000.0f * 421.8f)
                    radius:size.height / 1000.0f * 16.6f
                startAngle:5.2f * M_PI / 3.0f
                  endAngle:2.0f * M_PI / 3.0f
                 clockwise:YES];
    [path addLineToPoint:CGPointMake(size.width / 1000.0f * 146.0f, size.height / 1000.0f * 290.0f)];
    [path addLineToPoint:CGPointMake(size.width / 1000.0f * 130.9f, size.height / 1000.0f * 320.0f)];
    [path closePath];
    [path fill];

    // TODO: rework sickle
    // draw sickle
    [path removeAllPoints];
    [path moveToPoint:CGPointMake(size.width / 1000.0f * 169.0f, size.height / 1000.0f * 188.0f)];
    [path addCurveToPoint:CGPointMake(size.width / 1000.0f * 141.7f, size.height / 1000.0f * 377.3f)
            controlPoint1:CGPointMake(size.width / 1000.0f * 263.0f, size.height / 1000.0f * 260.5f)
            controlPoint2:CGPointMake(size.width / 1000.0f * 209.9f, size.height / 1000.0f * 492.0f)];
    [path addCurveToPoint:CGPointMake(size.width / 1000.0f * 132.2f, size.height / 1000.0f * 393.2f)
            controlPoint1:CGPointMake(size.width / 1000.0f * 138.1f, size.height / 1000.0f * 390.0f)
            controlPoint2:CGPointMake(size.width / 1000.0f * 138.0f, size.height / 1000.0f * 385.0f)];
    [path addCurveToPoint:CGPointMake(size.width / 1000.0f * 109.0f, size.height / 1000.0f * 437.0f)
            controlPoint1:CGPointMake(size.width / 1000.0f * 125.1f, size.height / 1000.0f * 410.0f)
            controlPoint2:CGPointMake(size.width / 1000.0f * 124.0f, size.height / 1000.0f * 439.0f)];
    [path addCurveToPoint:CGPointMake(size.width / 1000.0f * 105.1f, size.height / 1000.0f * 430.0f)
            controlPoint1:CGPointMake(size.width / 1000.0f * 106.0f, size.height / 1000.0f * 436.0f)
            controlPoint2:CGPointMake(size.width / 1000.0f * 106.0f, size.height / 1000.0f * 436.0f)];
    [path addCurveToPoint:CGPointMake(size.width / 1000.0f * 130.0f, size.height / 1000.0f * 370.0f)
            controlPoint1:CGPointMake(size.width / 1000.0f * 107.0f, size.height / 1000.0f * 387.0f)
            controlPoint2:CGPointMake(size.width / 1000.0f * 125.0f, size.height / 1000.0f * 400.0f)];
    [path addCurveToPoint:CGPointMake(size.width / 1000.0f * 141.0f, size.height / 1000.0f * 359.0f)
            controlPoint1:CGPointMake(size.width / 1000.0f * 133.0f, size.height / 1000.0f * 360.0f)
            controlPoint2:CGPointMake(size.width / 1000.0f * 135.0f, size.height / 1000.0f * 358.0f)];
    [path addLineToPoint:CGPointMake(size.width / 1000.0f * 144.5f, size.height / 1000.0f * 352.5f)];
    [path addCurveToPoint:CGPointMake(size.width / 1000.0f * 169.0f, size.height / 1000.0f * 188.0f)
            controlPoint1:CGPointMake(size.width / 1000.0f * 195.0f, size.height / 1000.0f * 435.0f)
            controlPoint2:CGPointMake(size.width / 1000.0f * 235.0f, size.height / 1000.0f * 270.0f)];
    [path closePath];
    [path fill];
    
    UIImage* image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return image;
}

@end

File: Sources/CountriesFlags/Countries/UIImage+SovietUnion.swift
//
//  UIImage+SovietUnion.swift
//  CountriesFlags
//

#if canImport(UIKit)

import UIKit

extension UIImage {
    /**
     [wikipedia]: https://en.wikipedia.org/wiki/Flag_of_the_Soviet_Union "wikipedia"
     
     Generates the flag of the Union of Soviet Socialist Republics.
     For more information see [wikipedia].
     
     - Parameters:
        - size: image size.
     
     - Returns: Returns the flag of the Union of Soviet Socialist Republics.
     */
    internal static func flagSovietUnion(with size: CGSize) -> UIImage? {
        let colorRed = UIColor(red: 203.0 / 255.0,
                               green: 0.0,
                               blue: 0.0,
                               alpha: 1.0)
        let colorYellow = UIColor(red: 1.0,
                                  green: 215.0 / 255.0,
                                  blue: 0.0,
                                  alpha: 1.0)

        UIGraphicsBeginImageContextWithOptions(size, false, 0.0)

        colorRed.setFill()
        UIRectFill(CGRect(origin: .zero, size: size))

        // TODO: rework star
        colorYellow.setFill()
//        colorYellow.setStroke()
        let starCenter = CGPoint(x: size.width / 100.0 * 16.7, y: size.height / 100.0 * 12.5)
        UIBezierPath.star(at: starCenter, radius: size.height / 16.0).fill()
        colorRed.setFill()
        UIBezierPath.star(at: starCenter, radius: size.height / 24.5).fill()

        colorYellow.setFill()
//        colorYellow.setStroke()
        // draw hammer
        let path = UIBezierPath()
        path.move(to: CGPoint(x: size.width / 1000.0 * 116.0, y: size.height / 1000.0 * 290.0))
        path.addLine(to: CGPoint(x: size.width / 1000.0 * 148.0, y: size.height / 1000.0 * 227.0))
        path.addLine(to: CGPoint(x: size.width / 1000.0 * 175.0, y: size.height / 1000.0 * 234.0))
        path.addLine(to: CGPoint(x: size.width / 1000.0 * 155.3, y: size.height / 1000.0 * 273.0))
        path.addLine(to: CGPoint(x: size.width / 1000.0 * 225.0, y: size.height / 1000.0 * 406.0))
        path.addArc(withCenter: CGPoint(x: size.width / 1000.0 * 221.0, y: size.height / 1000.0 * 421.8),
                    radius: size.height / 1000.0 * 16.6,
                    startAngle: 5.2 * .pi / 3.0,
                    endAngle: 2.0 * .pi / 3.0,
                    clockwise: true)
        path.addLine(to: CGPoint(x: size.width / 1000.0 * 146.0, y: size.height / 1000.0 * 290.0))
        path.addLine(to: CGPoint(x: size.width / 1000.0 * 130.9, y: size.height / 1000.0 * 320.0))
        path.close()
        path.fill()

        // TODO: rework sickle
        // draw sickle
        path.removeAllPoints()
        path.move(to: CGPoint(x: size.width / 1000.0 * 169.0, y: size.height / 1000.0 * 188.0))
        path.addCurve(to: CGPoint(x: size.width / 1000.0 * 141.7, y: size.height / 1000.0 * 377.3),
                      controlPoint1: CGPoint(x: size.width / 1000.0 * 263.0, y: size.height / 1000.0 * 260.5),
                      controlPoint2: CGPoint(x: size.width / 1000.0 * 209.9, y: size.height / 1000.0 * 492.0))
        path.addCurve(to: CGPoint(x: size.width / 1000.0 * 132.2, y: size.height / 1000.0 * 393.2), // handle
                      controlPoint1: CGPoint(x: size.width / 1000.0 * 138.1, y: size.height / 1000.0 * 390.0),
                      controlPoint2: CGPoint(x: size.width / 1000.0 * 138.0, y: size.height / 1000.0 * 385.0))
        path.addCurve(to: CGPoint(x: size.width / 1000.0 * 109.0, y: size.height / 1000.0 * 437.0),
                      controlPoint1: CGPoint(x: size.width / 1000.0 * 125.1, y: size.height / 1000.0 * 410.0),
                      controlPoint2: CGPoint(x: size.width / 1000.0 * 124.0, y: size.height / 1000.0 * 439.0))
        path.addCurve(to: CGPoint(x: size.width / 1000.0 * 105.1, y: size.height / 1000.0 * 430.0), // bottom
                      controlPoint1: CGPoint(x: size.width / 1000.0 * 106.0, y: size.height / 1000.0 * 436.0),
                      controlPoint2: CGPoint(x: size.width / 1000.0 * 106.0, y: size.height / 1000.0 * 436.0))
        path.addCurve(to: CGPoint(x: size.width / 1000.0 * 130.0, y: size.height / 1000.0 * 370.0),
                      controlPoint1: CGPoint(x: size.width / 1000.0 * 107.0, y: size.height / 1000.0 * 387.0),
                      controlPoint2: CGPoint(x: size.width / 1000.0 * 125.0, y: size.height / 1000.0 * 400.0))
        path.addCurve(to: CGPoint(x: size.width / 1000.0 * 141.0, y: size.height / 1000.0 * 359.0),
                      controlPoint1: CGPoint(x: size.width / 1000.0 * 133.0, y: size.height / 1000.0 * 360.0),
                      controlPoint2: CGPoint(x: size.width / 1000.0 * 135.0, y: size.height / 1000.0 * 358.0))
        path.addLine(to: CGPoint(x: size.width / 1000.0 * 144.5, y: size.height / 1000.0 * 352.5))
        path.addCurve(to: CGPoint(x: size.width / 1000.0 * 169.0, y: size.height / 1000.0 * 188.0),
                      controlPoint1: CGPoint(x: size.width / 1000.0 * 195.0, y: size.height / 1000.0 * 435.0),
                      controlPoint2: CGPoint(x: size.width / 1000.0 * 235.0, y: size.height / 1000.0 * 270.0))
        path.close()
        path.fill()
//        path.stroke()

        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()

        return image
    }
}

#endif
